libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx.h

```diff
@@ -678,6 +678,369 @@ _LIBCUDACXX_DEVICE static inline void st_async(
+      asm volatile (
+        "fence.sc.sys; // 1."
+        :
+        :
+        : "memory"
+      );
+    } else if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (__sem == sem_acq_rel && __scope == scope_cta) {
+      asm volatile (
+        "fence.acq_rel.cta; // 1."
+        :
+        :
+        : "memory"
+      );
+    } else if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (__sem == sem_acq_rel && __scope == scope_gpu) {
+      asm volatile (
+        "fence.acq_rel.gpu; // 1."
+        :
+        :
+        : "memory"
+      );
+    } else if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (__sem == sem_acq_rel && __scope == scope_sys) {
+      asm volatile (
+        "fence.acq_rel.sys; // 1."
+        :
+        :
+        : "memory"
+      );
+    }
+
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    __cuda_ptx_fence_is_not_supported_before_SM_70__();
+    return;
+  ));
+}
+#endif // __cccl_ptx_isa >= 600
+
+/*
+// fence{.sem}.scope; // 2. PTX ISA 78, SM_90
+// .sem       = { .sc, .acq_rel }
+// .scope     = { .cluster }
+template <cuda::ptx::dot_sem Sem>
+__device__ static inline void fence(
+  cuda::ptx::sem_t<Sem> sem,
+  cuda::ptx::scope_cluster_t);
+*/
+#if __cccl_ptx_isa >= 780
+extern "C" _LIBCUDACXX_DEVICE void __cuda_ptx_fence_is_not_supported_before_SM_90__();
+template <dot_sem _Sem>
+_LIBCUDACXX_DEVICE static inline void fence(
+  sem_t<_Sem> __sem,
+  scope_cluster_t)
+{
+  static_assert(__sem == sem_sc || __sem == sem_acq_rel, "");
+  // __scope == scope_cluster (due to parameter type constraint)
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,(
+    if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (__sem == sem_sc) {
+      asm volatile (
+        "fence.sc.cluster; // 2."
+        :
+        :
+        : "memory"
+      );
+    } else if _LIBCUDACXX_CONSTEXPR_AFTER_CXX14 (__sem == sem_acq_rel) {
+      asm volatile (
+        "fence.acq_rel.cluster; // 2."
+        :
+        :
+        : "memory"
+      );
+    }
+
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    __cuda_ptx_fence_is_not_supported_before_SM_90__();
+    return;
+  ));
+}
+#endif // __cccl_ptx_isa >= 780
+/*
+// fence.mbarrier_init.sem.scope; // 3. PTX ISA 80, SM_90
+// .sem       = { .release }
+// .scope     = { .cluster }
+template <typename=void>
+__device__ static inline void fence_mbarrier_init(
+  cuda::ptx::sem_release_t,
+  cuda::ptx::scope_cluster_t);
+*/
+#if __cccl_ptx_isa >= 800
+extern "C" _LIBCUDACXX_DEVICE void __cuda_ptx_fence_mbarrier_init_is_not_supported_before_SM_90__();
+template <typename=void>
+_LIBCUDACXX_DEVICE static inline void fence_mbarrier_init(
+  sem_release_t,
+  scope_cluster_t)
+{
+  // __sem == sem_release (due to parameter type constraint)
+  // __scope == scope_cluster (due to parameter type constraint)
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,(
+    asm volatile (
+      "fence.mbarrier_init.release.cluster; // 3."
+      :
+      :
+      : "memory"
+    );
+
```

Comment:
Question There are a lot of empty lines at the end of the NV_IF_TARGET block. Should we remove them?
```suggestion
```

Reply:
I removed the whitespace at the end of the `NV_IF_TARGET` block.